src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs

```diff
@@ -600,21 +600,22 @@ internal void InitializeForReflectionSerializer()
         private JsonTypeInfo GetJsonTypeInfoFromContextOrCreate(Type type)
         {
             JsonTypeInfo? info = _serializerContext?.GetTypeInfo(type);
-            if (info != null)
+            if (info == null && IsInitializedForReflectionSerializer)
             {
-                return info;
+                Debug.Assert(
+                    s_typeInfoCreationFunc != null,
+                    "Reflection-based JsonTypeInfo creator should be initialized if IsInitializedForReflectionSerializer is true.");
+                info = s_typeInfoCreationFunc(type, this);
             }
 
-            if (!IsInitializedForReflectionSerializer)
+            if (info == null)
             {
                 ThrowHelper.ThrowNotSupportedException_NoMetadataForType(type);
                 return null!;
             }
 
-            Debug.Assert(
-                s_typeInfoCreationFunc != null,
-                "Reflection-based JsonTypeInfo creator should be initialized if IsInitializedForReflectionSerializer is true.");
-            return s_typeInfoCreationFunc(type, this);
+            info.EnsureConfigured();
```

Comment:
If we're calling `EnsureConfigured` here, do we really need to repeat the call in the serialization entrypoints? I suspect this means we could move `EnsureConfigured` responsibility to the func itself.

Reply:
yes because serialization is also hit through some APIs which directly pass in JsonType info

Reply:
Would it make sense to rename the method to something like `EnsureInitialized()` or `EnsureLocked()` to better communicate what it does to the instance?

Reply:
actually in my original prototype I called it "EnsureLocked" but here it's not really locking anything yet. What about I rename it to "EnsureLocked" once it starts ensuring no one tries to modify it?

Reply:
> but here it's not really locking anything yet

I'm guessing though that it would eventually lock the public setters we'd be adding in the future? I think it's ok to introduce the name now even though it doesn't actually do exactly that since it signifies intent. Alternatively "EnsureInitialized" can be a generic enough description.

My issue with "EnsureConfigured" is that it's not really configuring anything (that happens on construction), it's just building caches based on configuration that has already been specified. Again though, that's me nit-picking on internal method names.

Reply:
What about BuildCache/EnsureCacheBuilt. I get similar notion with EnsureInitialized as you're getting with EnsureConfigured

Reply:
I think `EnsureInitialized` is good enough and adequately communicates intent.